src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,131 @@
+/**
+ * @file batch_norm_impl.hpp
+ * @author Praveen Ch
+ * @author Manthan-R-Sheth
+ *
+ * Implementation of the Batch Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+
+// In case it is not included.
+#include "batch_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm() :
+    size(10),
+    eps(1e-7),
+    deterministic(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm(
+    const size_t size, const double eps) :
+    size(size),
+    eps(eps),
+    deterministic(false)
+{
+  weights.set_size(size + size, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void BatchNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+  deterministic = false;
+  gamma.fill(1.0);
+  beta.fill(0.0);
+  stats.reset();
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BatchNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  output.reshape(input.n_rows, input.n_cols);
+
+  // Mean and variance over the entire training set will be used to compute
+  // the forward pass when deterministic is set to true.
+  if (deterministic)
+  {
+    mean = stats.mean();
+    variance = stats.var(1);
+  }
+  else
+  {
+    mean = arma::mean(input, 1);
+    variance = arma::var(input, 1, 1);
+
+    for (size_t i = 0; i < output.n_cols; i++)
+    {
+      stats(input.col(i));
+    }
+  }
+
+  output = input.each_col() - mean;
+  output.each_col() %= gamma / arma::sqrt(variance+eps);
```

Comment:
Picky style comment, use a space between operations; that also applies for other places e.g. line 94. Also, do you think we should write  ``gamma / (arma::sqrt(variance) + eps)`` here?

Reply:
arma::sqrt(variance +eps) will be closer to variance as compared to (arma::sqrt(variance) + eps) for varainace > 1. For variance < 1, the latter will be closer but the difference will be very less only.
Technically both are fine, but I think going with first one is better as it will be closer to true variance when variance > 1 (which will be the case most of the times).